diff --git a/src/slate/editor_ui_state.cljs b/src/slate/editor_ui_state.cljs
index 734b1be..bfaa28c 100644
--- a/src/slate/editor_ui_state.cljs
+++ b/src/slate/editor_ui_state.cljs
@@ -1,29 +1,29 @@
 (ns slate.editor-ui-state
   (:require-macros [slate.interceptors :refer [definterceptor]])
   (:require [clojure.set :as set]
-            [clojure.string :as str]
             [clojure.spec.alpha :as s]
+            [clojure.string :as str]
             [drop.utils :as utils]
-            [slate.model.dll :as dll]
+            [slate.default-interceptors :refer [default-interceptors]]
+            [slate.interceptors :as interceptors]
+            [slate.measurement :refer [ruler-for-elem]]
             [slate.model.common :as m]
+            [slate.model.dll :as dll :refer [big-dec]]
+            [slate.model.doc]
+            [slate.model.editor-state :as es]
             [slate.model.find-and-replace :as f+r]
             [slate.model.history :as history]
-            [slate.model.editor-state :as es]
-            [slate.model.doc]
             [slate.model.paragraph]
             [slate.model.run]
+            [slate.model.run :as r]
             [slate.model.selection :as sel]
-            [slate.interceptors :as interceptors]
-            [slate.default-interceptors :refer [default-interceptors]]
-            [slate.measurement :refer [ruler-for-elem]]
             [slate.renderer.core :as renderer]
-            [slate.serialization :refer [serialize deserialize]]
-            [slate.view :as view]
-            [slate.viewmodel :as vm]
+            [slate.serialization :refer [deserialize serialize]]
             [slate.style :as style]
-            [slate.word-count :as word-count]
             [slate.utils :as slate-utils]
-            [slate.model.run :as r]))
+            [slate.view :as view]
+            [slate.viewmodel :as vm]
+            [slate.word-count :as word-count]))
 
 (s/def ::id uuid?)
 (s/def ::history ::history/editor-state-history)
@@ -652,24 +652,27 @@
     (set! (.. shadow-dom-wrapper -shadowRoot -innerHTML)
           (str "<style id='" style-elem-id "'>" (style/get-rendered-shadow-elem-css font-family dark-mode?) "</style>"))
 
-    ;; There are some things you cannot do (like set outerHTML on elements, among other
-    ;; general weirdness) if an element is the immediate child of a <html> or ShadowRoot,
-    ;; so a top-level wrapper element is desirable over inserting straight into the shadow DOM.
-    #_(let [editor-elem (doto (js/document.createElement "div")
-                          (.. -classList (add "slate-editor")))]
-        (.. shadow-dom-wrapper -shadowRoot (appendChild editor-elem))
-        [editor-elem, (.-shadowRoot shadow-dom-wrapper)])
-    (let [width 600
+    (let [layers (doto (js/document.createElement "div")
+                   (aset "class" "layers"))
+          width 600
           height 400
           dpr js/window.devicePixelRatio
-          canvas-elem (doto (js/document.createElement "canvas")
-                        (aset "width" (* width dpr))
-                        (aset "height" (* height dpr))
-                        (aset "style" "width" (str width "px"))
-                        (aset "style" "height" (str height "px"))
-                        (.. -classList (add "slate-canvas")))]
-      (.. shadow-dom-wrapper -shadowRoot (appendChild canvas-elem))
-      [canvas-elem, (.-shadowRoot shadow-dom-wrapper)])))
+          text-layer-canvas (doto (js/document.createElement "canvas")
+                              (aset "width" (* width dpr))
+                              (aset "height" (* height dpr))
+                              (aset "style" "width" (str width "px"))
+                              (aset "style" "height" (str height "px"))
+                              (.. -classList (add "slate-canvas" "text-layer")))
+          caret-layer-canvas (doto (js/document.createElement "canvas")
+                               (aset "width" (* width dpr))
+                               (aset "height" (* height dpr))
+                               (aset "style" "width" (str width "px"))
+                               (aset "style" "height" (str height "px"))
+                               (.. -classList (add "slate-canvas" "caret-layer")))]
+      (.appendChild layers text-layer-canvas)
+      (.appendChild layers caret-layer-canvas)
+      (.. shadow-dom-wrapper -shadowRoot (appendChild layers))
+      [{:text text-layer-canvas, :caret caret-layer-canvas}, (.-shadowRoot shadow-dom-wrapper)])))
 
 (defn load-fonts!
   "Returns a Promise that resolves when the necessary fonts for rendering the document are loaded."
@@ -753,10 +756,11 @@
       (then #(let [uuid (random-uuid)
                    dark-mode? (= theme :dark)
                    ;; Slate operates inside a shadow DOM to prevent global styles from interfering
-                   [canvas-elem, shadow-root] (init-shadow-dom! dom-elem font-family dark-mode?)
+                   [canvases, shadow-root] (init-shadow-dom! dom-elem font-family dark-mode?)
                    #_#_available-width (.-width (.getBoundingClientRect (.-host shadow-root)))
                    #_#_measure-fn (ruler-for-elem editor-elem shadow-root)
-                   editor-state (es/editor-state)
+                   ;; editor-state (es/editor-state)
+                   editor-state (es/editor-state sample-doc (sel/selection [(big-dec 1) 82]))
                    history (history/init editor-state)
                    interceptors-map (-> (interceptors/interceptor-map)
                                         (interceptors/reg-interceptors default-interceptors)
@@ -793,7 +797,7 @@
                               :on-selection-changed on-selection-changed
                               :should-lose-focus? should-lose-focus?
                               :ready? false})
-               (renderer/init! canvas-elem sample-doc font-family 16 25)
+               (renderer/init! canvases editor-state font-family 16 25)
                #_(init-event-handlers! *atom)
                #_(full-dom-render! *atom)
                (swap! *atom assoc :ready? true)
diff --git a/src/slate/renderer/bst.cljs b/src/slate/renderer/bst.cljs
index 0ab17a2..1d15779 100644
--- a/src/slate/renderer/bst.cljs
+++ b/src/slate/renderer/bst.cljs
@@ -23,8 +23,8 @@
 
 (set! *warn-on-infer* false)
 
-(defn lt [a b] (.lt a b))
-(defn gt [a b] (.gt a b))
+(defn lt [a b] (if b (.lt a b) false))
+(defn gt [a b] (if b (.gt a b) false))
 
 (deftype AVLNode [^:mutable index
                   ^:mutable viewmodel
@@ -245,7 +245,9 @@
 
 (defn search
   [tree index]
-  (.-viewmodel (node-search (.-root tree) index)))
+  (when-let [node (node-search (.-root tree) index)]
+    ;; TODO: this is a hack. See also vm-at-y.
+    (assoc (.-viewmodel node) :y (.-height-above node))))
 
 (defn delete!
   [tree index]
@@ -256,7 +258,7 @@
 (defn vm-at-y
   [tree target-y-offset]
   (when-let [node (node-at-y (.-root tree) target-y-offset 0)]
-    ;; TODO: this works via a hack. Make sumn more elegant
+    ;; TODO: this is a hack. Could be more elegant?
     (assoc (.-viewmodel node) :y (.-height-above node))))
 
 (defn first-vm
diff --git a/src/slate/renderer/core.cljs b/src/slate/renderer/core.cljs
index 55d3b9a..9561fb1 100644
--- a/src/slate/renderer/core.cljs
+++ b/src/slate/renderer/core.cljs
@@ -3,7 +3,8 @@
             [slate.renderer.bst :as bst]
             [slate.renderer.measurement :refer [get-measure-fn]]
             [slate.renderer.utils :refer [font-str]]
-            [slate.renderer.viewmodel :as vm]))
+            [slate.renderer.viewmodel :as vm]
+            [slate.model.selection :as sel]))
 
 ;; renderer operations:
 ;; insert
@@ -44,18 +45,22 @@
           (.fillText ctx substring current-x y)
           (recur (rest substrings) (+ current-x (.-width text-metrics))))))))
 
-(defn render-span!
+(defn draw-caret! [ctx x y line-height]
+  (set! (.-fillStyle ctx) "#0085f2")
+  (.fillRect ctx x y 4 line-height))
+
+(defn draw-span!
   [ctx span font-family font-size line-y]
   (aset ctx "font" (font-str font-size font-family))
   (tab-aware-fill-text! ctx (:text span) 0 line-y))
 
-(defn render-vm!
+(defn draw-vm!
   [ctx paragraph-vm paragraph-y font-family font-size]
   (loop [lines (:lines paragraph-vm)
          line-y paragraph-y]
     (when-let [line (first lines)]
       (doseq [span (:spans line)]
-        (render-span! ctx span font-family font-size line-y))
+        (draw-span! ctx span font-family font-size line-y))
       (recur (rest lines)
              (+ line-y (get-line-height ctx font-family font-size))))))
 
@@ -79,7 +84,33 @@
 
 (defprotocol IRenderer
   (scroll! [this delta-y])
-  (render! [this doc]))
+  (render-caret! [this editor-state])
+  (render-doc! [this doc])
+  (render! [this editor-state]))
+
+(defn split-span
+  "Splits the span into two at the paragraph offset, and return a vector of [before, after]."
+  [span offset]
+  (let [diff (- offset (:start-offset span))
+        before (.substring (:text span) 0 diff)
+        after (.substring (:text span) diff)]
+    [(assoc span :text before), (assoc span :text after)]))
+
+(defn spans-before-offset
+  "Returns all spans in the line before the given paragraph offset."
+  [line offset]
+  (reduce (fn [spans-before, span]
+            (let [span-end-offset (+ (count (:text span)) (:start-offset span))]
+              (cond
+                (<= span-end-offset offset)
+                (conj spans-before span)
+
+                (and (<= (:start-offset span) offset) (< offset span-end-offset))
+                (conj spans-before (nth (split-span span offset) 0))
+
+                :else
+                (reduced spans-before))))
+          [] (:spans line)))
 
 (deftype Renderer [bst
                    scroll-y
@@ -89,13 +120,36 @@
                    tab-size-px
                    font-family
                    base-font-size
-                   text-layer-ctx]
+                   measure-fn
+                   text-layer-ctx
+                   caret-layer-ctx]
   IRenderer
   (scroll! [this delta-y]
     (set! (.-scroll-y this) (min (max 0 (+ scroll-y delta-y)) (.-total-height-px bst))))
 
-  ;; Renders only what's currently in the viewport
-  (render! [_ doc]
+  (render-caret! [_ editor-state]
+    (.clearRect caret-layer-ctx 0 0 viewport-width-px viewport-height-px)
+    (let [vm (bst/search bst (sel/caret-para (:selection editor-state)))
+          selection (:selection editor-state)
+          lines (vec (:lines vm))
+          caret-offset (sel/caret selection)
+          line-idx (loop [i 0]
+                     (let [{:keys [start-offset end-offset]} (nth lines i)]
+                       (if (and (<= start-offset caret-offset)
+                                (> end-offset caret-offset))
+                         i
+                         (recur (inc i)))))
+          line (nth lines line-idx)
+          line-height (get line-heights (:paragraph-type vm))
+          screen-y (+ (- scroll-y (:y vm)) (* line-idx line-height))
+          spans (spans-before-offset line caret-offset)
+          screen-x (reduce (fn [x {:keys [text formats]}]
+                             (+ x (measure-fn text formats (:paragraph-type vm))))
+                           0 spans)]
+      (draw-caret! caret-layer-ctx screen-x screen-y (:body line-heights))))
+
+;; Renders only what's currently in the viewport
+  (render-doc! [_ doc]
     (.clearRect text-layer-ctx 0 0 viewport-width-px viewport-height-px)
     (let [bottom-y (+ scroll-y viewport-height-px)
           first-visible-vm (first-visible-viewmodel bst scroll-y)
@@ -106,37 +160,45 @@
                                      (if last-visible-vm
                                        (:paragraph-index last-visible-vm)
                                        (dll/last-index (:children doc))))
-             paragraph-y (- (:body line-heights) first-visible-vm-offset)]
+             paragraph-y (+ (get line-heights (:paragraph-type first-visible-vm)) first-visible-vm-offset)] ; Here?
         (when-let [idx (first idxs)]
           (let [vm (bst/search bst idx)]
-            (render-vm! text-layer-ctx vm paragraph-y font-family base-font-size)
-            (recur (rest idxs) (+ paragraph-y (* (count (:lines vm)) (:body line-heights))))))))))
+            (draw-vm! text-layer-ctx vm paragraph-y font-family base-font-size)
+            (recur (rest idxs) (+ paragraph-y (* (count (:lines vm)) (:body line-heights)))))))))
+
+  (render! [this editor-state]
+    (render-doc! this (:doc editor-state))
+    (render-caret! this editor-state)))
 
 (defn init!
   "Initializes the Slate canvas renderer and does the initial render."
-  [canvas doc font-family base-font-size tab-size-px]
-  (let [dpr js/window.devicePixelRatio
-        width (.-width canvas)
-        height (* dpr (.-height canvas))
-        ctx (doto (.getContext canvas "2d")
-              #_(set-canvas-dimensions!))
-        body-line-height (get-line-height ctx font-family base-font-size)
+  [canvases editor-state font-family base-font-size tab-size-px]
+  (let [text-canvas (:text canvases)
+        caret-canvas (:caret canvases)
+        dpr js/window.devicePixelRatio
+        width (.-width text-canvas)
+        height (* dpr (.-height text-canvas))
+        text-ctx (.getContext text-canvas "2d")
+        caret-ctx (.getContext caret-canvas "2d")
+        body-line-height (get-line-height text-ctx font-family base-font-size)
         line-heights {:body body-line-height}
         measure-fn (get-measure-fn font-family base-font-size tab-size-px)
-        bst (init-bst doc width measure-fn line-heights)
-        renderer (Renderer. bst 0 width height line-heights tab-size-px font-family base-font-size ctx)]
+        bst (init-bst (:doc editor-state) width measure-fn line-heights)
+        renderer (Renderer. bst
+                            0
+                            width
+                            height
+                            line-heights
+                            tab-size-px
+                            font-family
+                            base-font-size
+                            measure-fn
+                            text-ctx
+                            caret-ctx)]
     (.addEventListener js/document "wheel" (fn [e]
                                              (scroll! renderer (.-deltaY e))
-                                             (render! renderer doc)))
-
-    ;; #p (bst/vm-at-y bst 64)
-    ;; (debug-tree/debug (.-root bst))
-    ;; (bst/traverse-in-order (.-root bst) (fn [n]
-    ;;                                      (js/console.log  (str (.-index n) " : " (bst/height-px n)))))
-    ;; (let [hello-width (measure-fn "Hello" #{} :body)]
-    ;;   (.fillRect ctx hello-width 0 100 100)
-    ;;   (.fillRect ctx 1173 0 100 100))
-    (render! renderer doc)
+                                             (render! renderer editor-state)))
+    (render! renderer editor-state)
     renderer))
 
 
diff --git a/src/slate/style.cljs b/src/slate/style.cljs
index ea5aa51..87d1ecc 100644
--- a/src/slate/style.cljs
+++ b/src/slate/style.cljs
@@ -65,7 +65,8 @@
                     :user-select "none"
                     :color "var(--text-color)"}
     [:&:hover {:cursor "text"}]]
-   [:.slate-canvas {:white-space "pre"
+   [:.slate-canvas {:position "absolute"
+                    :white-space "pre"
                     :box-sizing "border-box"
                     :tab-size "25px"
                     :margin 0
